### PR TITLE
Feature/metadata callback

### PR DESCRIPTION
**Pull request by sarah-segel**
Add an example for using a callback to log run metadata to a file. 
Logging this information will be important for the SMAC Data Collection (#996).

### CODE REVIEW

**Review by benjamc — CHANGES_REQUESTED** [review]
Could you put the metadata callback into `smac/callback/metadata_callback.py`?
This way we do not need to copy and paste the callback into our code but can just import it.

**Review by benjamc — CHANGES_REQUESTED** [review]
Rename folder `callbacks` to `callback` as this is the synthax for all other folders
After that we are ready to merge :)

**Comment by sarah-segel** [comment]
Renamed directory and moved `callback.py` there. Please re-check @benjamc :)

**Review by benjamc — APPROVED** [review]

File: smac/callback/metadata_callback.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+from smac.callback.callback import Callback
+from smac.main.smbo import SMBO
+
+__copyright__ = "Copyright 2023, AutoML.org Freiburg-Hannover"
+__license__ = "3-clause BSD"
+
+
+class MetadataCallback(Callback):
+    def __init__(self, **kwargs: str) -> None:
```
[the remark this answers is not in the document]
Comment (benjamc) [reply smac/callback/metadata_callback.py:11]: I agree that we should make obvious that the inputs must be serializable. @sarah-segel  I would propose to update the signature as Eddie wrote

**Comment by benjamc** [comment]
> How does this interact with SMAC's ability to restart from it's last run?

Actually I hope this does not interact. If we restart from the last run and write into the same folder it is the same run and the metadata should not have changed.
I am not sure how to handle cases like: We run SMAC once, commit sth, and run it further.
Do you have any other ideas what other metadata might be useful when gathering SMAC outputs, @eddiebergman ?

**Comment by eddiebergman** [comment]
You could also collect information like opearting system, time, smac version? Might be excessive but think about what you might like if someone was to raise an issue and has this readily available to report to you.

**Comment by benjamc** [comment]
Agreed. I think this is valuable to add.

File: examples/4_advanced_optimizer/3_metadata_callback.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        model.train,
+        overwrite=True,
+        callbacks=[
+            MetadataCallback(
+                project_name="My Project Name",
+                repository="My Repository Name",
+                branch="Name of Active Branch",
+                commit="Commit Hash",
+                command=" ".join(sys.argv),
+                utc_time=str(datetime.utcnow()),
```
[the remark this answers is not in the document]
Comment (sarah-segel) [reply examples/4_advanced_optimizer/3_metadata_callback.py:71]: Would make sense for me - at least for time, smac-version, and platform I would say?

**Pull request merged**